### PR TITLE
lncli: allow openchannel to accept a node's URI

### PR DESCRIPTION
**Pull request by jamesob**
Many websites (recksplorer, lightninggem) advertise their node in terms of URI.
Allow `lncli openchannel` to accept a URI (pubkey@host:port) without throwing a
decode error, and use the hostname information to attempt a connection without
having to specify `--connect`.

This removes `--connect` as it is no longer needed.

### Example usage

```
 $ lncli openchannel --sat_per_byte 13 02ad6fb8d693dc1e4569bcedefadf5f72a931ae027dc0f0c544b34c1c6f3b9a02b@167.99.50.31:9735 6000000 0

{
        "funding_txid": "d780dd9ff1adc76dfa2f32d05e22d2101c448296b8741262a398c2e3xxxxx"
}
```

### CODE REVIEW

**Comment by jamesob** [comment]
@wpaulino Thanks for the look. Removing `--connect` sounds fine with me and I'm happy to do so, but I'm wondering how sensitive lncli is to maintaining backwards compatibility. Do we want to first deprecate this flag and then remove it in a subsequent release, or are we okay with an unceremonious removal?

**Comment by wpaulino** [comment]
I'm on board with removing it, it's a CLI flag so it would only impact those workflows. We usually recommend people to stay pinned to the latest tagged release if they're looking for stability anyway. Will leave the decision to the rest of the team.

**Comment by jamesob** [comment]
I've pushed changes removing `--connect` (in lieu of URI use) as suggested by @wpaulino.

File: cmd/lncli/commands.go [hunk 1]
```diff
@@ -495,32 +489,35 @@ func openChannel(ctx *cli.Context) error {
...
-		req.NodePubkey = nodePubHex
+		nodePubHex, connectVal, err = decodeNodePubkeyOrURI(
+			ctx.String("node_key"))
 
 	case args.Present():
-		nodePubHex, err := hex.DecodeString(args.First())
-		if err != nil {
-			return fmt.Errorf("unable to decode node public key: %v", err)
-		}
+		nodePubHex, connectVal, err = decodeNodePubkeyOrURI(
```
Comment (halseth) [inline cmd/lncli/commands.go]: nit: I believe `req.NodePubkey` can be used here directly.

File: cmd/lncli/commands.go [hunk 2]
```diff
@@ -628,6 +625,20 @@ func openChannel(ctx *cli.Context) error {
 	}
 }
 
+func decodeNodePubkeyOrURI(val string) ([]byte, string, error) {
+	var nodeKey, connectVal string
+	split := strings.Split(val, "@")
+
+	if len(split) > 1 {
+		nodeKey, connectVal = split[0], split[1]
```
Comment (halseth) [inline cmd/lncli/commands.go]: looks like `nodeKey` is never set in the case where no URI is given?

**Comment by halseth** [comment]
I'm in favour of removing `--connect` 👍

File: cmd/lncli/commands.go (hunk 1 above)
Comment (jamesob) [reply cmd/lncli/commands.go]: I don't think so given that the first arg (like `node_key` above) can either be a pubkey or a URI, and in the latter case we need to unpack it, but maybe I'm missing something.

File: cmd/lncli/commands.go (hunk 2 above)
Comment (jamesob) [reply cmd/lncli/commands.go]: D'oh! Good find. Willfix.

**Comment by jamesob** [comment]
Pushed changes incorporating @halseth's feedback.

**Comment by jamesob** [comment]
Tested with both URI and pubkey on mainnet:
```sh
$ lncli openchannel --sat_per_byte 40 024a2e265cd66066b78a788ae615acdc84b5b0dec9efac36d7ac87513015eaf6ed@lnd.bitrefill.com 9000000 0

{
        "funding_txid": "3fb66e74b8ece13eebf883d9d210adacc9390bcdd446c580bde59ffc80ab342d"
}

File: cmd/lncli/commands.go (hunk 1 above)
Comment (halseth) [reply cmd/lncli/commands.go]: to clarify, what I meant was the following 😄 Not sure if it works?

```go
	switch {
	case ctx.IsSet("node_key"):
		req.NodePubkey, connectVal, err = decodeNodePubkeyOrURI(
			ctx.String("node_key"))

	case args.Present():
		req.NodePubkey, connectVal, err = decodeNodePubkeyOrURI(
			args.First())
		args = args.Tail()
	default:
		return fmt.Errorf("node id argument missing")
	}
```

**Comment by Roasbeef** [comment]
Needs a rebase!

**Pull request closed without merging**